src/PureScript/Backend/Optimizer/Semantics.purs

```diff
@@ -196,6 +196,28 @@ addStop (Env env) ref acc = Env env
       env.directives
   }
 
+
+floatLetFromArr :: Env -> Array BackendSemantics -> BackendSemantics
+floatLetFromArr = go []
+  where
+  go acc env = Array.uncons >>> case _ of
+    Nothing -> guardFailOver identity (LitArray acc) NeutLit
+    Just { head, tail } -> evalAssocLet env head (\e v -> go (acc <> [ v ]) e tail)
+
+floatLetFromRec :: Env -> Array (Prop BackendSemantics) -> BackendSemantics
+floatLetFromRec = go []
+  where
+  go acc env = Array.uncons >>> case _ of
+    Nothing -> guardFailOver identity (LitRecord acc) NeutLit
+    Just { head: Prop k v, tail } -> evalAssocLet env v (\e x -> go (acc <> [ Prop k x ]) e tail)
+
+floatLetFromCtor :: Qualified Ident -> ConstructorType -> ProperName -> Ident -> Env -> Array (Tuple String BackendSemantics) -> BackendSemantics
+floatLetFromCtor qual ct ty tag = go []
+  where
+  go acc env = Array.uncons >>> case _ of
+    Nothing -> guardFailOver snd acc $ NeutData qual ct ty tag
+    Just { head: Tuple k v, tail } -> evalAssocLet env v (\e x -> go (acc <> [ Tuple k x ]) e tail)
```

Comment:
We are a little inconsistent wrt to recursion over List vs `foldr`, but I would prefer either of those to recursion over Array.uncons. Additionally, if these aren't reused then I think we could just inline these definitions.

https://github.com/aristanetworks/purescript-backend-optimizer/blob/main/src/PureScript/Backend/Optimizer/Semantics.purs#L315-L322
https://github.com/aristanetworks/purescript-backend-optimizer/blob/main/src/PureScript/Backend/Optimizer/Semantics.purs#L439-L452